Mathlib/RingTheory/PowerSeries/WeierstrassPreparation.lean

```diff
@@ -435,4 +454,191 @@ theorem weierstrassMod_zero_left [IsAdicComplete (IsLocalRing.maximalIdeal A) A]
+  g.IsWeierstrassFactorizationAt f h (IsLocalRing.maximalIdeal A)
+
+namespace IsWeierstrassFactorizationAt
+
+variable {g : A⟦X⟧} {f : A[X]} {h : A⟦X⟧} {I : Ideal A} (H : g.IsWeierstrassFactorizationAt f h I)
+include H
+
+theorem map_ne_zero_of_ne_top (hI : I ≠ ⊤) : g.map (Ideal.Quotient.mk I) ≠ 0 := by
+  rcases subsingleton_or_nontrivial (A ⧸ I) with h | _
+  · apply (Submodule.Quotient.subsingleton_iff _).1 at h
+    exact False.elim <| hI <| by ext; simp [h]
+  rw [congr(map (Ideal.Quotient.mk I) $(H.eq_mul)), map_mul, ← Polynomial.polynomial_map_coe, ne_eq,
+    (H.isUnit.map _).mul_left_eq_zero]
+  exact_mod_cast f.map_monic_ne_zero (f := Ideal.Quotient.mk I) H.isDistinguishedAt.monic
+
+theorem degree_eq_coe_lift_order_map_of_ne_top (hI : I ≠ ⊤) :
+    f.degree = (g.map (Ideal.Quotient.mk I)).order.lift
+      (order_finite_iff_ne_zero.2 (H.map_ne_zero_of_ne_top hI)) := by
+  refine H.isDistinguishedAt.degree_eq_coe_lift_order_map g h ?_ H.eq_mul
+  contrapose! hI
+  exact Ideal.eq_top_of_isUnit_mem _ hI (isUnit_iff_constantCoeff.1 H.isUnit)
+
+theorem natDegree_eq_toNat_order_map_of_ne_top (hI : I ≠ ⊤) :
+    f.natDegree = (g.map (Ideal.Quotient.mk I)).order.toNat := by
+  rw [Polynomial.natDegree, H.degree_eq_coe_lift_order_map_of_ne_top hI,
+    ENat.lift_eq_toNat_of_lt_top]
+  exact WithBot.unbotD_coe _ _
+
+end IsWeierstrassFactorizationAt
+
+variable [IsLocalRing A]
+
+namespace IsWeierstrassFactorization
+
+variable {g : A⟦X⟧} {f : A[X]} {h : A⟦X⟧} (H : g.IsWeierstrassFactorization f h)
+include H
+
+theorem map_ne_zero : g.map (IsLocalRing.residue A) ≠ 0 :=
+  H.map_ne_zero_of_ne_top (Ideal.IsMaximal.ne_top inferInstance)
+
+theorem degree_eq_coe_lift_order_map : f.degree = (g.map (IsLocalRing.residue A)).order.lift
+    (order_finite_iff_ne_zero.2 H.map_ne_zero) :=
+  H.degree_eq_coe_lift_order_map_of_ne_top (Ideal.IsMaximal.ne_top inferInstance)
+
+theorem natDegree_eq_toNat_order_map :
+    f.natDegree = (g.map (IsLocalRing.residue A)).order.toNat :=
+  H.natDegree_eq_toNat_order_map_of_ne_top (Ideal.IsMaximal.ne_top inferInstance)
+
+end IsWeierstrassFactorization
+
+theorem IsWeierstrassDivision.isUnit_of_map_ne_zero
+    {g q : A⟦X⟧} {r : A[X]} (hg : g.map (IsLocalRing.residue A) ≠ 0)
+    (H : (X ^ (g.map (IsLocalRing.residue A)).order.toNat).IsWeierstrassDivision g q r) :
+    IsUnit q := by
+  obtain ⟨H1 : r.degree < (g.map (IsLocalRing.residue A)).order.toNat, H2⟩ := H
+  set n := (g.map (IsLocalRing.residue A)).order.toNat
+  replace H2 := congr(coeff _ n (($H2).map (IsLocalRing.residue A)))
+  simp_rw [map_pow, map_X, coeff_X_pow_self, map_add, map_mul, coeff_map,
+    Polynomial.coeff_coe, Polynomial.coeff_eq_zero_of_degree_lt H1, map_zero, add_zero] at H2
+  rw [coeff_mul, ← Finset.sum_subset (s₁ := {(n, 0)}) (by simp) (fun p hp hnmem ↦ by
+    have hp1 : p.1 < n := (Finset.antidiagonal.fst_le hp).lt_of_ne <| by
+      contrapose! hnmem
+      rwa [Finset.mem_singleton, Finset.antidiagonal_congr hp (by simp)]
```

Comment:
I find this a bit convoluted, with *three* by-proofs within a single `rw`. IMHO it would be nicer to single what you need as a standalone piece.

Reply:
Let me try.